relayer/src/supervisor/scan.rs

```diff
@@ -0,0 +1,774 @@
+    pub counterparty_state: Option<ConnectionState>,
+    pub channels: BTreeMap<ChannelId, ChannelScan>,
+}
+
+impl ConnectionScan {
+    pub fn new(
+        connection: IdentifiedConnectionEnd,
+        counterparty_state: Option<ConnectionState>,
+    ) -> Self {
+        Self {
+            connection,
+            counterparty_state,
+            channels: BTreeMap::new(),
+        }
+    }
+
+    pub fn id(&self) -> &ConnectionId {
+        &self.connection.connection_id
+    }
+
+    pub fn state(&self) -> ConnectionState {
+        self.connection.connection_end.state
+    }
+
+    pub fn is_open(&self) -> bool {
+        self.connection.connection_end.is_open()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct ChannelScan {
+    pub channel: IdentifiedChannelEnd,
+    pub counterparty: Option<IdentifiedChannelEnd>,
+}
+
+impl ChannelScan {
+    pub fn new(channel: IdentifiedChannelEnd, counterparty: Option<IdentifiedChannelEnd>) -> Self {
+        Self {
+            channel,
+            counterparty,
+        }
+    }
+
+    pub fn id(&self) -> &ChannelId {
+        &self.channel.channel_id
+    }
+
+    pub fn unreceived_packets_on_counterparty(
+        &self,
+        chain: &impl ChainHandle,
+        counterparty_chain: &impl ChainHandle,
+    ) -> Option<Vec<u64>> {
+        self.counterparty.as_ref().map(|counterparty| {
+            unreceived_packets(counterparty_chain, chain, counterparty).unwrap_or_default()
+        })
+    }
+
+    pub fn unreceived_acknowledgements_on_counterparty(
+        &self,
+        chain: &impl ChainHandle,
+        counterparty_chain: &impl ChainHandle,
+    ) -> Option<Vec<u64>> {
+        self.counterparty.as_ref().map(|counterparty| {
+            unreceived_acknowledgements(counterparty_chain, chain, counterparty).unwrap_or_default()
+        })
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ScanMode {
+    Auto,
+    Full,
+}
+
+pub struct ChainScanner<'a, Chain: ChainHandle> {
+    config: &'a Config,
+    registry: &'a mut Registry<Chain>,
+    client_state_filter: &'a mut FilterPolicy,
+    scan_mode: ScanMode,
+}
+
+impl<'a, Chain: ChainHandle> ChainScanner<'a, Chain> {
+    pub fn new(
+        config: &'a Config,
+        registry: &'a mut Registry<Chain>,
+        client_state_filter: &'a mut FilterPolicy,
+        scan_mode: ScanMode,
+    ) -> Self {
+        Self {
+            config,
+            registry,
+            client_state_filter,
+            scan_mode,
+        }
+    }
+
+    pub fn scan_chains(mut self) -> ChainsScan {
+        let mut scans = ChainsScan {
+            chains: Vec::with_capacity(self.config.chains.len()),
+        };
+
+        for chain in self.config.chains.clone() {
+            scans.chains.push(self.scan_chain(&chain));
+        }
+
+        scans
+    }
+
+    pub fn scan_chain(&mut self, chain_config: &ChainConfig) -> Result<ChainScan, Error> {
+        let span = info_span!("scan.chain", chain = %chain_config.id, );
```

Comment:
This is right, let's settle on `chain` key to record chain IDs in the logs so that the messages can be consistently filtered by chain ID. Lots of places use `chain.id` which is more verbose and contains problematic syntax, as described above.

A convention on span IDs may be more loose; when working on #1491 with @adizere, we tended to avoid lengthy IDs there, too, preferring nested spans to prefixes where it was needed for clarity.

Reply:
a414523d